<!-- <DOCUMENT>
	<FILE>
		9153364126.html
	</FILE>
	<URL>
		http://fire2.scl.genome.ad.jp/kegg/kegg4.html
	</URL>
	<TITLE>
		Links to Pathway and Other Databases
	</TITLE>
	<GENRE>
		linklists
	</GENRE>
	<PLAINTEXT>
 Links to Pathway and Other Databases Links to Pathway and Other Databases Metabolic Pathways KEGG Metabolic Pathways PUMA - Metabolic Pathways ---> EMP - Enzymes and Metabolic Pathways WIT - Metabolic Reconstruction UM-BBD - Microbial Biocatalysis/Biodegradatation EcoCyc - E. coli Genes and Metabolism SoyBase - Soybean Metabolism Metalgen - Genes and Metabolism Boehringer Mannheim - Biochemical Pathways IUBMB-Nicholson Minimaps Enzymes, Compounds and Reactions LIGAND - Biochemical Compounds and Reactions ENZYME - Enzymes BRENDA - Comprehensive Enzyme Information System Worthington Enzyme Manual Klotho - Biochemical Compounds ChemFinder - Searching Chemicals ChemIDplus at NLM PROMISE - Prosthetic Groups and Metal Ions GlycoSuiteDB - Glycan Structure Database CarbBank - Complex Carbohydrate Structure Database LIPIDBANK for Web - Lipids WebElements - Periodic Table Regulatory Pathways KEGG Regulatory Pathways SPAD - Signal Transduction CSNDB - Cell Signaling Networks Yeast Pathways in MIPS Interactive Fly - Drosophila Genes GIF_DB - Drosophila Gene Interactions FlyNets - Drosophila Molecular Interactions ---> GeNet - Gene Networks Database HOX-Pro - Homeobox Genes Database Drosophila Genes in Development ---> Wnt Signaling Pathway TRANSPATH - Gene Regulatory Pathways Protein-Protein Interactions BRITE Database for Biomolecular Relations DIP - Database of Interacting Proteins BIND - Biomolecular Interaction Network Database Transcription Factors TRANSFAC - Transcription Factor Database RegulonDB - E. coli Transcriptional Regulation DBTBS - B. subtilis Transcription Factors SCPD - S. cerevisiae Promoter Database DPInteract - DNA binding proteins Gene Expression Pattern Axeldb - Xenopus laevis NEXTDB - Caenorhabditis elegans MAGEST - Halocynthia roretzi Nomenclature - General IUBMB Nomenclature IUPAC Nomenclature IUPHAR Receptor Nomenclature SWISS-PROT Documents Gene Ontology Nomenclature - Organism-specific Human (HUGO) Mouse (MGD) Rat (RATMAP) D. melanogaster (FlyBase) C. elegans Arabidopsis (TAIR) Plants (Mendel) S. cerevisiae (SGD) Nomenclature and Classification - Protein-specific [EC:1.1.1.-] Alcohol dehydrogenase [EC:2.7.1.-] Protein kinases (PKC) [EC:3.1.4.-] Phosphodiesterases [EC:3.2.1.-] Glycosyl hydrolases (CAZy) / (ExPASy) [EC:3.4.-.-] Peptidases (MEROPS) / (ExPASy) [EC:6.1.1.-] Aminoacyl-tRNA synthetases (AARSDB) Cytochrome P450 Metallothionein / (ExPASy) CD Molecules (PROW) / (ExPASy) Immunoglobulins and T-cell receptors (IMGT) Cytokines (dbCFC) Transport proteins Ion channels (ICN) G protein coupled receptors (GPCRDB) G protein coupled receptors (GCRDb) Olfactory receptors (ORDB) Eph and Eph receptors Nuclear receptors Nuclear receptors (NRR) Mitochondrial proteins Ribosomal proteins (ExPASy) Homeobox proteins / (ExPASy) Taxonomy NCBI Taxonomy Tree of Life UCMP Phylogeny Exhibit Ribosomal Database Project II Species 2000 Merck Manual ---> Complete Genomes and Analysis NCBI Complete Microbial Genomes TIGR Comprehensive Microbial Resource GIB - Genome Information Broker PIR Complete Genomes PEDANT MAGPIE ---> GOLD MBGD - Comparative Microbial Genome Database COG - Clusters of Orthologous Groups Last updated: April 14, 2003 [ KEGG | GenomeNet ] 
	</PLAINTEXT>
	<CONTENT>
-->
<html>

<head>
<meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
<title>Links to Pathway and Other Databases</title>
<link rel="stylesheet" href="/css/kegg.css" type="text/css">
</head>

<body bgcolor="#ffffff" link="#003399" vlink="#003399">
<table BORDER=0 CELLPADDING=1 CELLSPACING=10>
<tr>
  <th><a href="/kegg/"><img src="/Fig/kegg128.gif" align="bottom" border=0 alt="KEGG"></a></th>
  <th><h2>Links to Pathway and Other Databases</h2></th>
</tr>
</table>
<hr>

<table cellspacing=5 cellpadding=5>
<tr valign="top">
<td>

<h4>Metabolic Pathways</h4>
<ul>
<li><a href="http://www.genome.ad.jp/kegg/metabolism.html">KEGG Metabolic Pathways</a></li>
<!---
<li><a href="http://www.mcs.anl.gov/home/compbio/PUMA/Production/puma_graphics.html">PUMA - Metabolic Pathways</a></li>
--->
<li><a href="http://emp.mcs.anl.gov/">EMP - Enzymes and Metabolic Pathways</a></li>
<li><a href="http://wit.mcs.anl.gov/WIT2/">WIT - Metabolic Reconstruction</a></li>
<li><a href="http://umbbd.ahc.umn.edu/">UM-BBD - Microbial Biocatalysis/Biodegradatation</a></li>
<li><a href="http://www.ecocyc.org/">EcoCyc - <I>E. coli</I> Genes and Metabolism</a></li>
<li><a href="http://cgsc.biology.yale.edu/metab.html">SoyBase - Soybean Metabolism</a></li>
<li><a href="http://indigo.genetique.uvsq.fr">Metalgen - Genes and Metabolism</a></li>
<li><a href="http://www.expasy.ch/cgi-bin/search-biochem-index">Boehringer Mannheim - Biochemical Pathways</a></li>
<li><a href="http://www.tcd.ie/Biochemistry/IUBMB-Nicholson/">IUBMB-Nicholson Minimaps</a></li>
</ul>

<h4>Enzymes, Compounds and Reactions</h4>
<ul>
<li><a href="http://www.genome.ad.jp/ligand/">LIGAND - Biochemical Compounds and Reactions</a></li>
<li><a href="http://www.expasy.ch/enzyme/">ENZYME - Enzymes</a></li>
<li><a href="http://www.brenda.uni-koeln.de/">BRENDA - Comprehensive Enzyme Information System</a></li>
<li><a href="http://www.worthington-biochem.com/manual/manIndex.html">Worthington Enzyme Manual</a></li>
<li><a href="http://www.biocheminfo.org/klotho/">Klotho - Biochemical Compounds</a></li>
<li><a href="http://chemfinder.camsoft.com/">ChemFinder - Searching Chemicals</a></li>
<li><a href="http://chem.sis.nlm.nih.gov/chemidplus/">ChemIDplus at NLM</a></li>
<li><a href="http://metallo.scripps.edu/PROMISE/">PROMISE - Prosthetic Groups and Metal Ions</a></li>
<li><a href="http://www.glycosuite.com/">GlycoSuiteDB - Glycan Structure Database</a></li>
<li><!A HREF="http://128.192.9.29/carbbank/"><a href="http://bssv01.lancs.ac.uk/gig/pages/gag/carbbank.htm">CarbBank - Complex Carbohydrate Structure Database</a></li>
<li><a href="http://lipidbank.jp/">LIPIDBANK for Web - Lipids</a></li>
<li><a href="http://www.webelements.com/">WebElements - Periodic Table</a></li>
</ul>

<h4>Regulatory Pathways</h4>
<ul>
<li><a href="http://www.genome.ad.jp/kegg/regulation.html">KEGG Regulatory Pathways</a></li>
<li><a href="http://www.grt.kyushu-u.ac.jp/spad/">SPAD - Signal Transduction</a></li>
<li><a href="http://geo.nihs.go.jp/csndb/">CSNDB - Cell Signaling Networks</a></li>
<li><a href="http://www.mips.biochem.mpg.de/proj/yeast/pathways/index.html">Yeast Pathways in MIPS</a></li>
<li><a href="http://sdb.bio.purdue.edu/fly/aimain/1aahome.htm">Interactive Fly - <I>Drosophila</I> Genes</a></li>
<!---
<li><a href="http://gifts.univ-mrs.fr/index.html">GIF_DB - <I>Drosophila</I> Gene Interactions</a></li>
<li><a href="http://gifts.univ-mrs.fr/FlyNets/FlyNets_home_page.html">FlyNets - <I>Drosophila</I> Molecular Interactions</a></li>
--->
<li><a href="http://www.csa.ru/Inst/gorb_dep/inbios/genet/genet.htm">GeNet - Gene Networks Database</a></li>
<li><a href="http://www.iephb.nw.ru/labs/lab38/spirov/hox_pro/hox-pro00.html">HOX-Pro - Homeobox Genes Database</a></li>
<!---
<li><a href="http://www.iephb.ru/~spirov/genmorph.html"><I>Drosophila</I> Genes in Development</a></li>
--->
<li><a href="http://www.stanford.edu/~rnusse/wntwindow.html">Wnt Signaling Pathway</a></li>
<li><a href="http://transpath.gbf.de/">TRANSPATH - Gene Regulatory Pathways</a></li>
</ul>

<h4>Protein-Protein Interactions</h4>
<ul>
<li><a href="http://www.genome.ad.jp/brite/">BRITE Database for Biomolecular Relations</a></li>
<li><a href="http://dip.doe-mbi.ucla.edu/">DIP - Database of Interacting Proteins</a></li>
<li><a href="http://www.binddb.org/">BIND - Biomolecular Interaction Network Database</a></li>
</ul>

<h4>Transcription Factors</h4>
<ul>
<li><a href="http://transfac.gbf.de/TRANSFAC/index.html">TRANSFAC - Transcription Factor Database</a></li>
<li><a href="http://www.cifn.unam.mx/Computational_Genomics/regulondb/">RegulonDB - <I>E. coli</I> Transcriptional Regulation</a></li>
<li><a href="http://elmo.ims.u-tokyo.ac.jp/dbtbs/">DBTBS - <I>B. subtilis</I> Transcription Factors</a></li>
<li><a href="http://cgsigma.cshl.org/jian/">SCPD - <I>S. cerevisiae</I> Promoter Database</a></li>
<li><a href="http://arep.med.harvard.edu/dpinteract/">DPInteract - DNA binding proteins</a></li>
</ul>

<h4>Gene Expression Pattern</h4>
<ul>
<li><a href="http://www.dkfz-heidelberg.de/abt0135/axeldb.htm">Axeldb - <I>Xenopus laevis</I></a></li>
<li><a href="http://nematode.lab.nig.ac.jp/">NEXTDB - <I>Caenorhabditis elegans</I></a></li>
<li><a href="http://www.genome.ad.jp/magest/">MAGEST - <I>Halocynthia roretzi</I></a></li>
</ul>

</td>
<td>

<h4>Nomenclature - General</h4>
<ul>
<li><a href="http://www.chem.qmul.ac.uk/iubmb/">IUBMB Nomenclature</a></li>
<li><a href="http://www.chem.qmul.ac.uk/iupac/">IUPAC Nomenclature</a></li>
<li><a href="http://www.iuphar-db.org/iuphar-rd/">IUPHAR Receptor Nomenclature</a></li>
<li><a href="http://www.expasy.ch/sprot/sp-docu.html">SWISS-PROT Documents</a></li>
<li><a href="http://www.geneontology.org/">Gene Ontology</a></li>
</ul>

<h4>Nomenclature - Organism-specific</h4>
<ul>
<li><a href="http://www.gene.ucl.ac.uk/nomenclature/">Human (HUGO)</a></li>
<li><a href="http://www.informatics.jax.org/mgihome/nomen/">Mouse (MGD)</a></li>
<li><a href="http://rgnc.gen.gu.se/RGNChem.html">Rat (RATMAP)</a></li>
<li><a href="http://flybase.bio.indiana.edu/docs/nomenclature/lk/nomenclature.html"><I>D. melanogaster</I> (FlyBase)</a></li>
<li><a href="http://elegans.swmed.edu/Genome/Nomencl2001w.htm"><I>C. elegans</I></a></li>
<li><a href="http://www.arabidopsis.org/nomencl.html"><I>Arabidopsis</I> (TAIR)</a></li>
<li><a href="http://www.mendel.ac.uk/">Plants (Mendel)</a></li>
<li><a href="http://genome-www.stanford.edu/Saccharomyces/registry.html"><I>S. cerevisiae</I> (SGD)</a></li>
</ul>

<h4>Nomenclature and Classification - Protein-specific</h4>
<ul>
<li>[EC:1.1.1.-] <a href="http://www.gene.ucl.ac.uk/nomenclature/genefamily/ADH-2.shtml">Alcohol dehydrogenase</a></li>
<li>[EC:2.7.1.-] <a href="http://pkr.sdsc.edu/html/index.shtml">Protein kinases (PKC)</a></li>
<li>[EC:3.1.4.-] <a href="http://depts.washington.edu/pde/Nomenclature.html">Phosphodiesterases</a></li>
<li>[EC:3.2.1.-] <a href="http://afmb.cnrs-mrs.fr/CAZY/">Glycosyl hydrolases (CAZy)</a> /</li>
    <a href="http://www.expasy.ch/cgi-bin/lists?glycosid.txt">(ExPASy)</a>
<li>[EC:3.4.-.-] <a href="http://merops.iapc.bbsrc.ac.uk/">Peptidases (MEROPS)</a> /</li>
    <a href="http://www.expasy.ch/cgi-bin/lists?peptidas.txt">(ExPASy)</a>
<li>[EC:6.1.1.-] <a href="http://rose.man.poznan.pl/aars/index.html">Aminoacyl-tRNA synthetases (AARSDB)</a></li>
<li><a href="http://drnelson.utmem.edu/CytochromeP450.html">Cytochrome P450</a></li>
<li><a href="http://www.unizh.ch/~mtpage/classif.html">Metallothionein</a> /</li>
    <a href="http://www.expasy.ch/cgi-bin/lists?metallo.txt">(ExPASy)</a>
<li><a href="http://www.ncbi.nlm.nih.gov/PROW/">CD Molecules (PROW)</a> /</li>
    <a href="http://www.expasy.ch/cgi-bin/lists?cdlist.txt">(ExPASy)</a>
<li><a href="http://imgt.cines.fr/">Immunoglobulins and T-cell receptors (IMGT)</a></li>
<li><a href="http://cytokine.medic.kumamoto-u.ac.jp/">Cytokines (dbCFC)</a></li>
<li><a href="http://tcdb.ucsd.edu/tcdb/">Transport proteins</a></li>
<li><a href="http://www.pain.med.umn.edu/csn/">Ion channels (ICN)</a></li>
<li><a href="http://www.gpcr.org/7tm/">G protein coupled receptors (GPCRDB)</a></li>
<li><a href="http://www.gcrdb.uthscsa.edu/">G protein coupled receptors (GCRDb)</a></li>
<li><a href="http://senselab.med.yale.edu/senselab/ORDB/">Olfactory receptors (ORDB)</a></li>
<li><a href="http://cbweb.med.harvard.edu/eph-nomenclature/">Eph and Eph receptors</a></li>
<li><a href="http://www.ens-lyon.fr/LBMC/laudet/nomenc.html">Nuclear receptors</a></li>
<li><a href="http://nrr.georgetown.edu/NRR/NRR.html">Nuclear receptors (NRR)</a></li>
<li><a href="http://mips.gsf.de/proj/medgen/mitop/">Mitochondrial proteins</a></li>
<li><a href="http://www.expasy.ch/cgi-bin/lists?ribosomp.txt">Ribosomal proteins (ExPASy)</a></li>
<li><a href="http://www.ncbi.nlm.nih.gov/entrez/query.fcgi?cmd=Retrieve&db=PubMed&list_uids=8098522&dopt=Abstract">Homeobox proteins</a> /</li>
    <a href="http://www.expasy.ch/cgi-bin/lists?hoxlist.txt">(ExPASy)</a>
</ul>

<h4>Taxonomy</h4>
<ul>
<li><a href="http://www.ncbi.nlm.nih.gov/Taxonomy/">NCBI Taxonomy</a></li>
<li><a href="http://tolweb.org/tree/phylogeny.html">Tree of Life</a></li>
<li><a href="http://www.ucmp.berkeley.edu/exhibit/phylogeny.html">UCMP Phylogeny Exhibit</a></li>
<li><a href="http://rdp.cme.msu.edu/html/">Ribosomal Database Project II</a></li>
<!---
<li><a href="http://www.sp2000.org/aboutsp2000.html">Species 2000</a></li>
<li><a href="http://www.merck.com/pubs/mmanual/sections.htm">Merck Manual</a></li>
--->
</ul>

<h4>Complete Genomes and Analysis</h4>
<ul>
<li><a href="http://www.ncbi.nlm.nih.gov/PMGifs/Genomes/micr.html">NCBI Complete Microbial Genomes</a></li>
<li><a href="http://www.tigr.org/tigr-scripts/CMR2/CMRHomePage.spl">TIGR Comprehensive Microbial Resource</a></li>
<li><a href="http://gib.genes.nig.ac.jp/">GIB - Genome Information Broker</a></li>
<li><a href="http://www-nbrf.georgetown.edu/pir/genome.html">PIR Complete Genomes</a></li>
<li><a href="http://pedant.mips.biochem.mpg.de/">PEDANT</a></li>
<!---
<li><a href="http://www-fp.mcs.anl.gov/~gaasterland/genomes.html">MAGPIE</a></li>
--->
<li><a href="http://ergo.integratedgenomics.com/GOLD/">GOLD</a></li>
<li><a href="http://mbgd.genome.ad.jp/">MBGD - Comparative Microbial Genome Database</a></li>
<li><a href="http://www.ncbi.nlm.nih.gov/COG/">COG - Clusters of Orthologous Groups</a></li>
</ul>

</td>
</tr>
</table>

<br>
<i>Last updated: April 14, 2003</i>
<hr>

<b>
<b>
[ <a href="/kegg/">KEGG</a> |
<a href="/">GenomeNet</a> ]
</b>

</body>

</html>

